// @threeceelabs/mongoose-twitter/index.js
const chalk = require("chalk");
const chalkInfo = chalk.gray;
const chalkAlert = chalk.red;
const chalkError = chalk.bold.red;
const chalkLog = chalk.bold.black;

const config = { wordAssoDb: "mongodb://127.0.0.1:27017/wordAsso?replicaSet=rs0" };
const mongoose = require("mongoose");
mongoose.Promise = global.Promise;

const tweetModel = require("./models/tweet.server.model");
const hashtagModel = require("./models/hashtag.server.model");
const urlModel = require("./models/url.server.model");
const mediaModel = require("./models/media.server.model");
const locationModel = require("./models/location.server.model");
const placeModel = require("./models/place.server.model");
const userModel = require("./models/user.server.model");
const emojiModel = require("./models/emoji.server.model");
const wordModel = require("./models/word.server.model");
const networkInputsModel = require("./models/networkInputs.server.model");
const neuralNetworkModel = require("./models/neuralNetwork.server.model");

let options = { 
	useNewUrlParser: true,
	keepAlive: 120,
	autoReconnect: true,
	autoIndex: false,
	reconnectTries: Number.MAX_VALUE,
	socketTimeoutMS: 90000,
	connectTimeoutMS: 0,
	poolSize: 100
};

let db;

exports.connect = function(appname, callback) {

	options.appname = appname;

	console.log(chalkLog("*** CREATING MONGO DB CONNECTION | NAME: " + options.appname));

	mongoose.connect(config.wordAssoDb, options);

	db = mongoose.connection;

  db.on("error", function (err) {
		console.log(chalkError("*** MONGO DB CONNECT ERROR *** | NAME: " + options.appname + "\n" + err));
		return(callback("MONGO DB CONNECT ERROR | " + options.appname, null));
  });

  db.once("open", function () {
		console.log(chalk.bold.green("+++ MONGO DB CONNECTED | NAME: " + options.appname + " | " + config.wordAssoDb));
		if (callback !== undefined ) { callback(null, mongoose.connection); }
  });

		// if (err) {
		// 	console.log(chalkError("*** MONGO DB CONNECT ERROR *** | NAME: " + options.appname + "\n" + err));
		// 	return(callback("MONGO DB CONNECT ERROR | " + options.appname, null));
		// }

		// console.log(chalk.bold.green("+++ MONGO DB CONNECTED | NAME: " + options.appname + " | " + config.wordAssoDb));
		// if (callback !== undefined ) { callback(null, mongoose.connection); }


	// mongoose.connect(config.wordAssoDb, options, function(err){

	// 	if (err) {
	// 		console.log(chalkError("*** MONGO DB CONNECT ERROR *** | NAME: " + options.appname + "\n" + err));
	// 		return(callback("MONGO DB CONNECT ERROR | " + options.appname, null));
	// 	}

	// 	console.log(chalk.bold.green("+++ MONGO DB CONNECTED | NAME: " + options.appname + " | " + config.wordAssoDb));
	// 	if (callback !== undefined ) { callback(null, mongoose.connection); }

	// });

};
