// @threeceelabs/mongoose-twitter/models/neuralNetwork.server.model.js
const mongoose = require("mongoose");
const testCycleHistory = require("../plugins/testCycleHistory");

mongoose.Promise = global.Promise;

const Schema = mongoose.Schema;

exports.NeuralNetworkSchema = new Schema({
	networkFamily: { type: String	},
	networkId: { type: String, unique: true	},
	betterChild: { type: Boolean, default: false },
	seedNetworkId: { type: String	},
	seedNetworkRes: { type: String},
	networkCreateMode: { type: String	},
	createdAt: { type: Date, default: Date.now() },
	network: { type: mongoose.Schema.Types.Mixed,	default: {}	},
	successRate: { type: Number, default: 0	},
	matchRate: { type: Number, default: 0	},
	overallMatchRate: { type: Number, default: 0 },
	testCycles: { type: Number, default: 0 },
	testCycleHistory: { type: Array, default: [] },
	numInputs: { type: Number, default: 0	},
	numOutputs: { type: Number,	default: 0 },
	inputsId: { type: String	},
	inputsObj: { type: mongoose.Schema.Types.Mixed, default: {} },
	outputs: { type: mongoose.Schema.Types.Mixed, default: {}	},
	evolve: { type: mongoose.Schema.Types.Mixed,	default: {}	},
	train: { type: mongoose.Schema.Types.Mixed, default: {}	},
	test: {	type: mongoose.Schema.Types.Mixed, default: {} }
});

exports.NeuralNetworkSchema.plugin(testCycleHistory);
