// @threeceelabs/mongoose-twitter/models/place.server.model.js
const mongoose = require("mongoose");
mongoose.Promise = global.Promise;
const Schema = mongoose.Schema;

exports.PlaceSchema = new Schema({
	placeId: {
		type: String,
		unique: true
	},
	nodeType: { 
		type: String,
		default: "place"
	},
	nodeId: { 
		type: String
	},
	tags: { 
		type: Schema.Types.Mixed,
		default: {}
	},
	isKeyword: { 
		type: Boolean,
		default: true
	},
	keywords: {  
		type: Schema.Types.Mixed,
		default: { "neutral": 100}
	},
	keywordsAuto: {
		type: Schema.Types.Mixed,
		default: {}
	},
	geocodeNodeId: { 
		type: String
	},
	formattedAddress: { 
		type: String
	},
	boundingBox: { 
		type: Schema.Types.Mixed
	},
	centroid: { 
		type: [Number],
		default: []
	},
	name: { 
		type: String
	},
	fullName: { 
		type: String
	},
	countryCode: { 
		type: String
	},
	country: { 
		type: String
	},
	placeType: { 
		type: String
	},
	url: {   
		type: String
	},
	sourceUrl: {   
		type: String
	},
	imageUrl: {   
		type: String
	},
	createdAt: {   // in ms
		// type: String
		type: Date, default: Date.now
	},
	lastSeen: {   // in ms
		// type: String
		type: Date, default: Date.now
	},
	lastTweetId: { type: String },
	mentions: {
		type: Number,
		default: 0
	}
});
