// @threeceelabs/mongoose-twitter/models/tweet.server.model.js
const mongoose = require("mongoose");
mongoose.Promise = global.Promise;
const Schema = mongoose.Schema;

exports.TweetSchema = new Schema({
	tweetId: { 
		type: String,
		trim: true,
		unique: true
	},
	nodeId: {  // for D3.js
		type: String
	},
	nodeType: { 
		type: String,
		default: "tweet"
	},
	lang: { type: String },
	user: {
		type: mongoose.Schema.Types.Mixed
	},
	userMentions: {
		type: [mongoose.Schema.Types.Mixed],
		default: []
	},
	url: {
		type: String,
		trim: true
	},
	profileImageUrl: {
		type: String,
		trim: true
	},
	imageUrl: {
		type: String,
		trim: true
	},
	createdAt: {   // in ms
		// type: String
		type: Date, default: Date.now
	},
	lastSeen: {   // in ms
		// type: String
		type: Date, default: Date.now
	},
	mentions: {
		type: Number,
		default: 0
	},
	quotedId: {
		type: String,
		default: 0
	},
	quotedStatus: {
		type: mongoose.Schema.Types.Mixed
	},
	retweeted: {
		type: Boolean,
		default: false
	},
	isRetweet: {
		type: Boolean,
		default: false
	},
	retweetedId: {
		type: String,
		default: 0
	},
	retweetedStatus: {
		type: mongoose.Schema.Types.Mixed
	},
	retweets: {
		type: String,
		default: 0
	},
	favorites: {
		type: Number,
		default: 0
	},
	text: {
		type: String
	},
	extendedText: {
		type: String
	},
	lang: {
		type: String
	},
	translations: { // translation = { lang: "en", text: "What?"}
		type: mongoose.Schema.Types.Mixed
	},
	hashtags: {
		type: [mongoose.Schema.Types.Mixed],
		default: []
	},
	place: {
		type: mongoose.Schema.Types.Mixed,
		trim: true
	},
	media: {
		type: [mongoose.Schema.Types.Mixed],
		default: []
	},
	urls: {
		type: [mongoose.Schema.Types.Mixed],
		default: []
	},
	rate: {
		type: Number,
		default: 0
	},
	status: mongoose.Schema.Types.Mixed
});

// mongoose.model("Tweet", TweetSchema);
