// @threeceelabs/mongoose-twitter/models/user.server.model.js
const mongoose = require("mongoose");
const moment = require("moment");
mongoose.Promise = global.Promise;
const Schema = mongoose.Schema;

exports.UserSchema = new Schema({
	oauthID: { type: Number },
	userId: { type: String},
	nodeId: { type: String, unique: true },
	nodeType: { type: String, default: "user"},
	lang: { type: String },

	name: { type: String, default: "" },
	previousName: { type: String, default: "" },

	description: { type: String, default: "" },
	previousDescription: { type: String, default: "" },

	isTwitterUser: { type: Boolean },
	screenName: { type: String },
	screenNameLower: { type: String },
	url: { type: String	},
	profileUrl: { type: String },
	profileImageUrl: { type: String },
	bannerImageUrl: { type: String },
	verified: { type: Boolean, default: false },
	location: { type: String },

	createdAt: { type: Date, default: Date.now() },
	lastSeen: { type: Date, default: Date.now() },

	status: { type: mongoose.Schema.Types.Mixed, default: {} },
	statusId: { type: String, default: "0" }, // last tweet id sent by user
	previousStatusId: { type: String, default: "0" }, // last tweet id sent by user

	lastTweetId: { type: String, default: "0" }, // either last tweet id sent by user OR tweet where user is mentioned

	mentions: { type: Number, default: 0 },
	
	statusesCount: { type: Number, default: 0 },
	followersCount: { type: Number, default: 0 },

	friendsCount: { type: Number, default: 0 },
	friends: { type: mongoose.Schema.Types.Mixed, default: [] },

	countHistory: { type: mongoose.Schema.Types.Mixed, default: [] },

	ignored: { type: Boolean, default: false },
	following: { type: Boolean,	default: false },
	threeceeFollowing: { type: String },

	rate: { type: Number, default: 0 },
	isTopTerm: { type: Boolean, default: false },

	category: { type: mongoose.Schema.Types.Mixed, default: false },
	categoryAuto: { type: mongoose.Schema.Types.Mixed, default: false },
	
	keywords: { type: mongoose.Schema.Types.Mixed, default: false },
	keywordsAuto: { type: mongoose.Schema.Types.Mixed, default: false },

	bannerImageAnalyzed: { type: mongoose.Schema.Types.Mixed, default: false },

	histograms: { type: mongoose.Schema.Types.Mixed, default: false },
	lastHistogramTweetId: { type: String, default: "0" },

	languageAnalyzed: { type: Boolean, default: false },
	languageAnalysis: { type: mongoose.Schema.Types.Mixed, default: {} }
});
