// @threeceelabs/mongoose-twitter/models/word.server.model.js
const mongoose = require("mongoose");
// const moment = require("moment");
mongoose.Promise = global.Promise;
const Schema = mongoose.Schema;

exports.WordSchema = new Schema({
	nodeId: { 
		type: String,
		unique: true
	},
	nodeType: { type: String, default: "word" },
	raw: { 
		type: String
	},
	url: { 
		type: String
	},
	createdAt: {   // in ms
		// type: String
		type: Date, default: Date.now
	},
	lastSeen: {   // in ms
		// type: String
		type: Date, default: Date.now
	},
	lastTweetId: { type: String },
	sessionId: {   
		type: String
	},
	wordChainIndex: {   
		type: Number,
		default: 0
	},
	links: {   
		type: mongoose.Schema.Types.Mixed
	},
	mentions: {
		type: Number,
		default: 0
	},
	rate: {
		type: Number,
		default: 0
	},
	isTwitterUser: {
		type: Boolean,
		default: false
	},
	isTopTerm: {
		type: Boolean,
		default: false
	},
	isTrendingTopic: {
		type: Boolean,
		default: false
	},
	isKeyword: {
		type: Boolean,
		default: false
	},
	category: { type: String },
	keywords: { type: mongoose.Schema.Types.Mixed, default: false },
	keywordsAuto: { type: mongoose.Schema.Types.Mixed, default: false },
	isIgnored: {
		type: Boolean,
		default: false
	},
	tags: {
		type: mongoose.Schema.Types.Mixed,
		default: {}
	},
	noun: {
		type: mongoose.Schema.Types.Mixed
	},
	verb: {
		type: mongoose.Schema.Types.Mixed
	},
	adjective: {
		type: mongoose.Schema.Types.Mixed
	},
	adverb: {
		type: mongoose.Schema.Types.Mixed
	},
	wapiSearched: {
		type: Boolean,
		default: false
	},
	wapiFound: {
		type: Boolean,
		default: false
	},
	wapiResults: {
		type: mongoose.Schema.Types.Mixed
	},
	bhtSearched: {
		type: Boolean,
		default: false
	},
	bhtFound: {
		type: Boolean,
		default: false
	},
	bhtAlt: {  // 303 Redirected
		type: String
	},
	mwDictSearched: {
		type: Boolean,
		default: false
	},
	mwDictFound: {
		type: Boolean,
		default: false
	},
	mwThesSearched: {
		type: Boolean,
		default: false
	},
	mwThesFound: {
		type: Boolean,
		default: false
	},
	mwEntry: {  // mw thesaurus
		type: mongoose.Schema.Types.Mixed
	},
	mwSuggestion: { // mw dictionary
		type: mongoose.Schema.Types.Mixed
	}
});
