// @threeceelabs/twitter-text-parser/index.js
/*jslint node: true */
"use strict";

/* TWITTER TEXT PARSER */
/* tweet-text-parser */

const async = require("async");
const debug = require("debug")("ttp");

const mentionsRegex = require("mentions-regex");
const hashtagRegex = require("hashtag-regex");
const getUrls = require("get-urls");

const chalk = require("chalk");
const chalkError = chalk.bold.red;
const chalkAlert = chalk.red;
const chalkWarn = chalk.red;
const chalkLog = chalk.gray;
const chalkInfo = chalk.black;

const emoji = require("node-emoji");

const emojiRegex = require("emoji-regex");
const eRegex = emojiRegex();
let Autolinker = require("autolinker");
const keywordExtractor = require("keyword-extractor");

const wordExtractionOptions = {
  language:"english",
  remove_digits: true,
  return_changed_case: true,
  remove_duplicates: true
};

let parser = new Autolinker({
  email: false,
  urls: true,
  hashtag: "twitter",
  mention: "twitter"
});

let histograms = {};
histograms.words = {};
histograms.urls = {};
histograms.hashtags = {};
histograms.mentions = {};
histograms.emoji = {};

const jsonPrint = function (obj){
  if (obj) {
    return JSON.stringify(obj, null, 2);
  }
  else {
    return "UNDEFINED";
  }
};

exports.clearGlobalHistograms = function (){
  histograms = {};
  histograms.words = {};
  histograms.urls = {};
  histograms.hashtags = {};
  histograms.mentions = {};
  histograms.emoji = {};
  histograms.images = {};
};

exports.getGlobalHistograms = function (callback){
  callback(histograms);
};

exports.parse = function (tweet, options, callback){

  // debug(chalk.blue("\nPARSE TEXT\n" + text + "\n"));
  debug(chalk.blue("\nPARSE OPTIONS\n" + jsonPrint(options) + "\n"));

  if (tweet === "undefined") {
    console.error(chalkError("*** PARSER TWEET UNDEFINED"));
  }

  const userHistograms = {};
  userHistograms.words = {};
  userHistograms.urls = {};
  userHistograms.hashtags = {};
  userHistograms.mentions = {};
  userHistograms.emoji = {};
  userHistograms.images = {};

  histograms.images = {};

  let comboTweetText = tweet.text + " " + tweet.user.screen_name;

  if (tweet.user.name !== undefined) {
    comboTweetText = comboTweetText + " " + tweet.user.name;
  }
  if (tweet.user.description !== undefined) {
    comboTweetText = comboTweetText + " " + tweet.user.description;
  }
  if (tweet.retweeted_status !== undefined) {
    comboTweetText = comboTweetText + " " + tweet.retweeted_status.text;
  }

  comboTweetText = comboTweetText.replace(/,/gi, " ");

  let matchEmoji;
  let emojiArray = [];

  while (matchEmoji = eRegex.exec(comboTweetText)) {

    const emj = matchEmoji[0];
    const unEmoji = emoji.unemojify(emj); // image -> :emoji:

    emojiArray.push(unEmoji);

    comboTweetText = comboTweetText.replace(emj, " ");

    debug(chalkAlert(emojiArray.length + " | EMJ: " + emj + " | " + unEmoji));

    debug(chalk.bold.black("\nTEXT LESS EMOJI: " + comboTweetText + "\n\n"));
  }

  let textWordBreaks = comboTweetText.replace(/\//gim, " ");
  const wordArray = keywordExtractor.extract(comboTweetText, wordExtractionOptions);

  async.parallel({

    mentions: function(cb){

      if (tweet.entities.user_mentions.length > 0) {

        async.each(tweet.entities.user_mentions, function(userObj, cb1){

          const screenName = userObj.screen_name.toLowerCase();

          if (options.updateGlobalHistograms) {

            if (histograms.mentions[screenName] === undefined) {
              histograms.mentions[screenName] = {};
              histograms.mentions[screenName].total = 0;
              histograms.mentions[screenName].left = 0;
              histograms.mentions[screenName].neutral = 0;
              histograms.mentions[screenName].right = 0;
              histograms.mentions[screenName].positive = 0;
              histograms.mentions[screenName].negative = 0;
              histograms.mentions[screenName].uncategorized = 0;
            }

            histograms.mentions[screenName].total += 1;

            if (options.category) {
              if (options.category === "left") { histograms.mentions[screenName].left += 1; }
              if (options.category === "neutral") { histograms.mentions[screenName].neutral += 1; }
              if (options.category === "right") { histograms.mentions[screenName].right += 1; }
              if (options.category === "positive") { histograms.mentions[screenName].positive += 1; }
              if (options.category === "negative") { histograms.mentions[screenName].negative += 1; }
            }
            else {
              histograms.mentions[screenName].uncategorized += 1;
            }
          }

          userHistograms.mentions[screenName] = (userHistograms.mentions[screenName] === undefined) ? 1 
            : userHistograms.mentions[screenName]+1;

          debug(chalkAlert("->- DESC Ms"
            + " | " + userHistograms.mentions[screenName]
            + " | @" + screenName
          ));

          cb1();

        }, function(err){
          cb(err, userHistograms.mentions);
        });
      }
      else {
        cb(null, userHistograms.mentions);
      }
    },

    hashtags: function(cb){

      if (tweet.entities.hashtags.length > 0) {

        async.each(tweet.entities.hashtags, function(hashtagObj, cb1){

          const hashtag = hashtagObj.text.toLowerCase();

          if (options.updateGlobalHistograms) {
            if (histograms.hashtags[hashtag] === undefined) {
              histograms.hashtags[hashtag] = {};
              histograms.hashtags[hashtag].total = 0;
              histograms.hashtags[hashtag].left = 0;
              histograms.hashtags[hashtag].neutral = 0;
              histograms.hashtags[hashtag].right = 0;
              histograms.hashtags[hashtag].positive = 0;
              histograms.hashtags[hashtag].negative = 0;
              histograms.hashtags[hashtag].uncategorized = 0;
            }

            histograms.hashtags[hashtag].total += 1;

            if (options.category) {
              if (options.category === "left") { histograms.hashtags[hashtag].left += 1; }
              if (options.category === "neutral") { histograms.hashtags[hashtag].neutral += 1; }
              if (options.category === "right") { histograms.hashtags[hashtag].right += 1; }
              if (options.category === "positive") { histograms.hashtags[hashtag].positive += 1; }
              if (options.category === "negative") { histograms.hashtags[hashtag].negative += 1; }
            }
            else {
              histograms.hashtags[hashtag].uncategorized += 1;
            }
          }

          userHistograms.hashtags[hashtag] = (userHistograms.hashtags[hashtag] === undefined) ? 1 
            : userHistograms.hashtags[hashtag]+1;

          debug(chalkAlert("->- DESC Hs"
            + " | " + userHistograms.hashtags[hashtag]
            + " | " + hashtag
          ));

          cb1();
        }, function(err){
          cb(err, userHistograms.hashtags);
        });
      }
      else {
        cb(null, userHistograms.hashtags);
      }
    },

    urls: function(cb){

      if (tweet.entities.urls.length > 0) {

        async.each(tweet.entities.urls, function(urlObj, cb1){

          const url = urlObj.expanded_url.replace(/(http:\/\/(www)?)|(https:\/\/(www)?)|[\.\-_\?&=\+\$\/#]/g, "").toLowerCase(); 
          debug(chalkAlert("URL REPLACE | " + url + " < " + urlObj.expanded_url));

          if (options.updateGlobalHistograms) {

            if (histograms.urls[url] === undefined) {
              histograms.urls[url] = {};
              histograms.urls[url].total = 0;
              histograms.urls[url].left = 0;
              histograms.urls[url].neutral = 0;
              histograms.urls[url].right = 0;
              histograms.urls[url].positive = 0;
              histograms.urls[url].negative = 0;
              histograms.urls[url].uncategorized = 0;
            }

            histograms.urls[url].total += 1;
            
            if (options.category) {
              if (options.category === "left") { histograms.urls[url].left += 1; }
              if (options.category === "neutral") { histograms.urls[url].neutral += 1; }
              if (options.category === "right") { histograms.urls[url].right += 1; }
              if (options.category === "positive") { histograms.urls[url].positive += 1; }
              if (options.category === "negative") { histograms.urls[url].negative += 1; }
            }
            else {
              histograms.urls[url].uncategorized += 1;
            }

          }

          userHistograms.urls[url] = (userHistograms.urls[url] === undefined) ? 1 : userHistograms.urls[url]+1;

          debug(chalkAlert("->- DESC Us"
            + " | " + userHistograms.urls[url]
            + " | " + url
          ));

          cb1();
        }, function(err){
          cb(err, userHistograms.urls);
        });
      }
      else {
        cb(null, userHistograms.urls);
      }
    },

    images: function(cb){

      if (tweet.extended_entities === undefined) {
        return cb(null, userHistograms.images);
      }

      async.each(tweet.extended_entities.media, function(mediaObj, cb1){

        const mediaId = mediaObj.id_str;

        if (options.updateGlobalHistograms) {

          if (histograms.images[mediaId] === undefined) {
            histograms.images[mediaId] = {};
            histograms.images[mediaId].total = 0;
            histograms.images[mediaId].left = 0;
            histograms.images[mediaId].neutral = 0;
            histograms.images[mediaId].right = 0;
            histograms.images[mediaId].positive = 0;
            histograms.images[mediaId].negative = 0;
            histograms.images[mediaId].uncategorized = 0;
          }

          histograms.images[mediaId].total += 1;
          
          if (options.category) {
            if (options.category === "left") { histograms.images[mediaId].left += 1; }
            if (options.category === "neutral") { histograms.images[mediaId].neutral += 1; }
            if (options.category === "right") { histograms.images[mediaId].right += 1; }
            if (options.category === "positive") { histograms.images[mediaId].positive += 1; }
            if (options.category === "negative") { histograms.images[mediaId].negative += 1; }
          }
          else {
            histograms.images[mediaId].uncategorized += 1;
          }

        }

        userHistograms.images[mediaId] = (userHistograms.images[mediaId] === undefined) ? 1 : userHistograms.images[mediaId]+1;

        debug(chalkAlert("->- DESC Us"
          + " | " + userHistograms.images[mediaId]
          + " | " + mediaId
        ));

        cb1();
      }, function(err){
        cb(err, userHistograms.images);
      });
    },

    words: function(cb){
      if (wordArray) {
        async.each(wordArray, function(w, cb1){

          debug(chalkAlert("w"
            + " | " + w
          ));

          let word = w.toLowerCase().trim();

          word = word.replace(/\$/gi, "");
          word = word.replace(/'s/gi, "");
          word = word.replace(/’s/gi, "");
          word = word.replace(/'ve/gi, "");
          word = word.replace(/’ve/gi, "");
          word = word.replace(/'re/gi, "");
          word = word.replace(/’re/gi, ""); 
          word = word.replace(/…/gi, ""); 

          const m = mentionsRegex().exec(word);
          const h = hashtagRegex().exec(word);

          let u;

          try {
            const urlObj = getUrls(word);
            u = (Array.from(urlObj).length > 0) ? Array.from(urlObj) : null;
          }
          catch (e) {
            console.error(chalkError("GET URLS ERROR\n" + jsonPrint(e)));
          }

          if (m || h || u 
            || (word === "/") 
            || word.includes("\u2026") 
            || word.includes("--") 
            || word.includes("|") 
            || word.includes("#") 
            || word.includes("…") 
            || word.includes("w/") 
            || word.includes("≠") 
            || word.includes("http") 
            || word.includes("+")) {
            debug(chalkAlert("-- SKIP WORD"
              + " | M: " + m
              + " | H: " + h
              + " | U: " + u
              + " | " + word
            ));
            cb1();
          }
          else {
            if (options.updateGlobalHistograms) {

              if (histograms.words[word] === undefined) {
                histograms.words[word] = {};
                histograms.words[word].total = 0;
                histograms.words[word].left = 0;
                histograms.words[word].neutral = 0;
                histograms.words[word].right = 0;
                histograms.words[word].positive = 0;
                histograms.words[word].negative = 0;
                histograms.words[word].uncategorized = 0;
              }

              histograms.words[word].total += 1;
              
              if (options.category) {
                if (options.category === "left") { histograms.words[word].left += 1; }
                if (options.category === "neutral") { histograms.words[word].neutral += 1; }
                if (options.category === "right") { histograms.words[word].right += 1; }
                if (options.category === "positive") { histograms.words[word].positive += 1; }
                if (options.category === "negative") { histograms.words[word].negative += 1; }
              }
              else {
                histograms.words[word].uncategorized += 1;
              }

            }
            userHistograms.words[word] = (userHistograms.words[word] === undefined) ? 1 
              : userHistograms.words[word]+1;

            debug(chalkAlert("->- DESC Ws"
              // + " | HIST: " + histograms.words[word]
              + " | " + userHistograms.words[word]
              + " | " + word
            ));

            cb1();
          }
        }, function(err){
          cb(null, userHistograms.words);
        });
      }
      else {
        cb(null, userHistograms.words);
      }
    },

    emoji: function(cb){
      if (emojiArray) {
        async.each(emojiArray, function(emoji, cb2){
          if (options.updateGlobalHistograms) {

            if (histograms.emoji[emoji] === undefined) {
              histograms.emoji[emoji] = {};
              histograms.emoji[emoji].total = 0;
              histograms.emoji[emoji].left = 0;
              histograms.emoji[emoji].neutral = 0;
              histograms.emoji[emoji].right = 0;
              histograms.emoji[emoji].positive = 0;
              histograms.emoji[emoji].negative = 0;
              histograms.emoji[emoji].uncategorized = 0;
            }

            histograms.emoji[emoji].total += 1;
            
            if (options.category) {
              if (options.category === "left") { histograms.emoji[emoji].left += 1; }
              if (options.category === "neutral") { histograms.emoji[emoji].neutral += 1; }
              if (options.category === "right") { histograms.emoji[emoji].right += 1; }
              if (options.category === "positive") { histograms.emoji[emoji].positive += 1; }
              if (options.category === "negative") { histograms.emoji[emoji].negative += 1; }
            }
            else {
              histograms.emoji[emoji].uncategorized += 1;
            }

          }
          userHistograms.emoji[emoji] = (userHistograms.emoji[emoji] === undefined) ? 1 : userHistograms.emoji[emoji]+1;
          debug(chalkAlert("->- DESC Es"
            + " | " + userHistograms.emoji[emoji]
            + " | " + emoji
          ));
          cb2();
        }, function(err2){
          cb(err2, userHistograms.emoji);
        });
      }
      else {
        cb(null, userHistograms.emoji);
      }
    }

  }, function(err, results){
    callback(err, results);
  });
};

exports.parseText = function (text, options, callback){

  debug(chalk.blue("\nPARSE TEXT\n" + text + "\n"));
  debug(chalk.blue("\nPARSE OPTIONS\n" + jsonPrint(options) + "\n"));

  if (text === "undefined") {
    console.error(chalkError("*** PARSER TEXT UNDEFINED"));
  }

  const userHistograms = {};
  userHistograms.words = {};
  userHistograms.urls = {};
  userHistograms.hashtags = {};
  userHistograms.mentions = {};
  userHistograms.emoji = {};
  histograms.images = {};

  text = text.replace(/,/gi, " ");

  const parseResults = parser.parse(text);

  let urlArray = [];
  let mentionArray = [];
  let hashtagArray = [];
  let emojiArray = [];


  async.each(parseResults, function(matchObj, cb){

    const type = matchObj.getType();

    debug(chalkAlert("PARSE TEXT"
      + " | " + matchObj.getMatchedText().toLowerCase()
      + " | TYPE: " + type
    ));

    // check for elipses
    if (/\u2026/.test(matchObj.getMatchedText())){
      console.log(chalkAlert(matchObj.getMatchedText().toLowerCase() + " CONTAINS ELIPSES ... SKIPPING ..."));
      return cb();
    }

    switch (type) {
      case "url":

        let url = matchObj.getMatchedText().replace("https://t.co/", "");  // MONGO DOESNT LIKE URLs
        url = url.replace(/https:\/\//gi, "");
        url = url.replace(/http:\/\//gi, "");
        url = url.replace(/\./gi, "_DOT_");

        urlArray.push(url.toLowerCase());
        debug(chalkInfo(urlArray.length + " | URL: " + url.toLowerCase()));
        cb();
      break;
      case "mention":
        mentionArray.push(matchObj.getMatchedText().toLowerCase());
        debug(chalkInfo(mentionArray.length + " | MEN: " + matchObj.getMatchedText().toLowerCase()));
        cb();
      break;
      case "hashtag":
        hashtagArray.push(matchObj.getMatchedText().toLowerCase());
        debug(chalkInfo(hashtagArray.length + " | HTG: " + matchObj.getMatchedText().toLowerCase()));
        cb();
      break;
      case "phone":
        cb();
      break;
      default:
        console.error(chalkError("UNKNOWN PARSE TYPE: " + type));
        cb();
    }

   }, function(err){

    let matchEmoji;

    while (matchEmoji = eRegex.exec(text)) {

      const emj = matchEmoji[0];
      const unEmoji = emoji.unemojify(emj); // image -> :emoji:

      emojiArray.push(unEmoji);

      text = text.replace(emj, " ");

      debug(chalkAlert(emojiArray.length + " | EMJ: " + emj + " | " + unEmoji));

      debug(chalk.bold.black("\nTEXT LESS EMOJI: " + text + "\n\n"));
    }
 
    let textWordBreaks = text.replace(/\//gim, " ");
    const wordArray = keywordExtractor.extract(text, wordExtractionOptions);

    async.parallel({

      mentions: function(cb){
        if (mentionArray) {

          async.each(mentionArray, function(userId, cb2){
            userId = userId.toLowerCase();
            if (options.updateGlobalHistograms) {
              if (histograms.mentions[userId] === undefined) {
                histograms.mentions[userId] = {};
                histograms.mentions[userId].total = 0;
                histograms.mentions[userId].left = 0;
                histograms.mentions[userId].neutral = 0;
                histograms.mentions[userId].right = 0;
                histograms.mentions[userId].positive = 0;
                histograms.mentions[userId].negative = 0;
                histograms.mentions[userId].uncategorized = 0;
              }

              histograms.mentions[userId].total += 1;
              if (options.category) {
                if (options.category === "left") { histograms.mentions[userId].left += 1; }
                if (options.category === "neutral") { histograms.mentions[userId].neutral += 1; }
                if (options.category === "right") { histograms.mentions[userId].right += 1; }
                if (options.category === "positive") { histograms.mentions[userId].positive += 1; }
                if (options.category === "negative") { histograms.mentions[userId].negative += 1; }
              }
              else {
                histograms.mentions[userId].uncategorized += 1;
              }
            }
            userHistograms.mentions[userId] = (userHistograms.mentions[userId] === undefined) ? 1 
              : userHistograms.mentions[userId]+1;
            debug(chalkAlert("->- DESC Ms"
              + " | " + userHistograms.mentions[userId]
              + " | " + userId
            ));
            cb2();
          }, function(err2){
            cb(err2, userHistograms.mentions);
          });
        }
        else {
          cb(null, userHistograms.mentions);
        }
      },

      hashtags: function(cb){
        if (hashtagArray) {
          async.each(hashtagArray, function(hashtag, cb2){
            hashtag = hashtag.toLowerCase();
            if (options.updateGlobalHistograms) {
              if (histograms.hashtags[hashtag] === undefined) {
                histograms.hashtags[hashtag] = {};
                histograms.hashtags[hashtag].total = 0;
                histograms.hashtags[hashtag].left = 0;
                histograms.hashtags[hashtag].neutral = 0;
                histograms.hashtags[hashtag].right = 0;
                histograms.hashtags[hashtag].positive = 0;
                histograms.hashtags[hashtag].negative = 0;
                histograms.hashtags[hashtag].uncategorized = 0;
              }

              histograms.hashtags[hashtag].total += 1;

              if (options.category) {
                if (options.category === "left") { histograms.hashtags[hashtag].left += 1; }
                if (options.category === "neutral") { histograms.hashtags[hashtag].neutral += 1; }
                if (options.category === "right") { histograms.hashtags[hashtag].right += 1; }
                if (options.category === "positive") { histograms.hashtags[hashtag].positive += 1; }
                if (options.category === "negative") { histograms.hashtags[hashtag].negative += 1; }
              }
              else {
                histograms.hashtags[hashtag].uncategorized += 1;
              }
            }

            userHistograms.hashtags[hashtag] = (userHistograms.hashtags[hashtag] === undefined) ? 1 
              : userHistograms.hashtags[hashtag]+1;
            debug(chalkAlert("->- DESC Hs"
              + " | " + userHistograms.hashtags[hashtag]
              + " | " + hashtag
            ));
            cb2();
          }, function(err2){
            cb(err2, userHistograms.hashtags);
          });
        }
        else {
          cb(null, userHistograms.hashtags);
        }
      },

      words: function(cb){

        if (wordArray) {

          async.each(wordArray, function(w, cb1){

            debug(chalkAlert("w"
              + " | " + w
            ));

            let word = w.toLowerCase().trim();

            word = word.replace(/\$/gi, "");
            word = word.replace(/'s/gi, "");
            word = word.replace(/’s/gi, "");
            word = word.replace(/'ve/gi, "");
            word = word.replace(/’ve/gi, "");
            word = word.replace(/'re/gi, "");
            word = word.replace(/’re/gi, "");
            word = word.replace(/…/gi, ""); 

            const m = mentionsRegex().exec(word);
            const h = hashtagRegex().exec(word);

            let u;

            try {
              const urlObj = getUrls(word);
              u = (Array.from(urlObj).length > 0) ? Array.from(urlObj) : null;
            }
            catch (e) {
              console.error(chalkError("GET URLS ERROR\n" + jsonPrint(e)));
            }

            if (m || h || u 
              || (word === "/") 
              || word.includes("\u2026") 
              || word.includes("--") 
              || word.includes("|") 
              || word.includes("#") 
              || word.includes("…") 
              || word.includes("w/") 
              || word.includes("≠") 
              || word.includes("http") 
              || word.includes("+")) {
              debug(chalkAlert("-- SKIP WORD"
                + " | M: " + m
                + " | H: " + h
                + " | U: " + u
                + " | " + word
              ));
              cb1();
            }
            else {
              if (options.updateGlobalHistograms) {

                if (histograms.words[word] === undefined) {
                  histograms.words[word] = {};
                  histograms.words[word].total = 0;
                  histograms.words[word].left = 0;
                  histograms.words[word].neutral = 0;
                  histograms.words[word].right = 0;
                  histograms.words[word].positive = 0;
                  histograms.words[word].negative = 0;
                  histograms.words[word].uncategorized = 0;
                }

                histograms.words[word].total += 1;
                
                if (options.category) {
                  if (options.category === "left") { histograms.words[word].left += 1; }
                  if (options.category === "neutral") { histograms.words[word].neutral += 1; }
                  if (options.category === "right") { histograms.words[word].right += 1; }
                  if (options.category === "positive") { histograms.words[word].positive += 1; }
                  if (options.category === "negative") { histograms.words[word].negative += 1; }
                }
                else {
                  histograms.words[word].uncategorized += 1;
                }

              }
              userHistograms.words[word] = (userHistograms.words[word] === undefined) ? 1 
                : userHistograms.words[word]+1;

              debug(chalkAlert("->- DESC Ws"
                // + " | HIST: " + histograms.words[word]
                + " | " + userHistograms.words[word]
                + " | " + word
              ));

              cb1();
            }
          }, function(err){
            cb(null, userHistograms.words);
          });
        }
        else {
          cb(null, userHistograms.words);
        }
      },

      urls: function(cb){
        if (urlArray) {
          async.each(urlArray, function(url, cb2){
            url = url.toLowerCase();
            if (options.updateGlobalHistograms) {

              if (histograms.urls[url] === undefined) {
                histograms.urls[url] = {};
                histograms.urls[url].total = 0;
                histograms.urls[url].left = 0;
                histograms.urls[url].neutral = 0;
                histograms.urls[url].right = 0;
                histograms.urls[url].positive = 0;
                histograms.urls[url].negative = 0;
                histograms.urls[url].uncategorized = 0;
              }

              histograms.urls[url].total += 1;
              
              if (options.category) {
                if (options.category === "left") { histograms.urls[url].left += 1; }
                if (options.category === "neutral") { histograms.urls[url].neutral += 1; }
                if (options.category === "right") { histograms.urls[url].right += 1; }
                if (options.category === "positive") { histograms.urls[url].positive += 1; }
                if (options.category === "negative") { histograms.urls[url].negative += 1; }
              }
              else {
                histograms.urls[url].uncategorized += 1;
              }

            }
            userHistograms.urls[url] = (userHistograms.urls[url] === undefined) ? 1 : userHistograms.urls[url]+1;
            debug(chalkAlert("->- DESC Us"
              + " | " + userHistograms.urls[url]
              + " | " + url
            ));
            cb2();
          }, function(err2){
            cb(err2, userHistograms.urls);
          });
        }
        else {
          cb(null, userHistograms.urls);
        }
      },

      emoji: function(cb){
        if (emojiArray) {
          async.each(emojiArray, function(emoji, cb2){
            if (options.updateGlobalHistograms) {

              if (histograms.emoji[emoji] === undefined) {
                histograms.emoji[emoji] = {};
                histograms.emoji[emoji].total = 0;
                histograms.emoji[emoji].left = 0;
                histograms.emoji[emoji].neutral = 0;
                histograms.emoji[emoji].right = 0;
                histograms.emoji[emoji].positive = 0;
                histograms.emoji[emoji].negative = 0;
                histograms.emoji[emoji].uncategorized = 0;
              }

              histograms.emoji[emoji].total += 1;
              
              if (options.category) {
                if (options.category === "left") { histograms.emoji[emoji].left += 1; }
                if (options.category === "neutral") { histograms.emoji[emoji].neutral += 1; }
                if (options.category === "right") { histograms.emoji[emoji].right += 1; }
                if (options.category === "positive") { histograms.emoji[emoji].positive += 1; }
                if (options.category === "negative") { histograms.emoji[emoji].negative += 1; }
              }
              else {
                histograms.emoji[emoji].uncategorized += 1;
              }

            }
            userHistograms.emoji[emoji] = (userHistograms.emoji[emoji] === undefined) ? 1 : userHistograms.emoji[emoji]+1;
            debug(chalkAlert("->- DESC Es"
              + " | " + userHistograms.emoji[emoji]
              + " | " + emoji
            ));
            cb2();
          }, function(err2){
            cb(err2, userHistograms.emoji);
          });
        }
        else {
          cb(null, userHistograms.emoji);
        }
      }

    }, function(err, results){
      callback(err, results);
    });

  });
};
